python/activator/middleware_interface.py

```diff
@@ -868,19 +919,19 @@ def _export_subset(self, exposure_ids: set[int],
         # TODO: get a proper synchronization API for Butler
         self.central_butler.registry.refresh()
 
-        with tempfile.NamedTemporaryFile(mode="w+b", suffix=".yaml") as export_file:
-            # MUST NOT export governor dimensions, as this causes deadlocks in
-            # central registry. Can omit most other dimensions (all dimensions,
-            # after DM-36051) to avoid locks or redundant work.
-            # TODO: saveDatasets(elements={"exposure", "visit"}) doesn't work.
-            # Use import(skip_dimensions) until DM-36062 is fixed.
-            with self.butler.export(filename=export_file.name) as export:
-                export.saveDatasets(datasets)
-            self.central_butler.import_(filename=export_file.name,
-                                        directory=self.butler.datastore.root,
-                                        skip_dimensions={"instrument", "detector",
-                                                         "skymap", "tract", "patch"},
-                                        transfer="copy")
+        # Transferring governor dimensions in parallel can cause deadlocks in
+        # central registry. We need to transfer our exposure/visit dimensions,
+        # so handle those manually.
+        for dimension in ["exposure", "visit"]:
+            for record in self.butler.registry.queryDimensionRecords(
```

Comment:
Does this work because we don't expect many exposure/visit records in the source butler? If there were a lot the alternative is to extract the relevant exposure/visit records directly from the refs that are being transferred.

Reply:
Right. I had missed that the where clause had a limited set of exposure IDs.

I don't really know what the data flow is here. Are you saying that the data turn up in prompt processing, the exposure record is first defined there, then you define the visits there, and then this is where all the results get copied to permanent storage? In that case it does seem like all the other visit dimensions will need to be included.

Reply:
(unless @TallJimbo tells me that something clever is going on inside syncDimensionData)

Reply:
bq. Are you saying that the data turn up in prompt processing, the exposure record is first defined there, then you define the visits there, and then this is where all the results get copied to permanent storage?

Yes, that's correct. The raw is first ingested into the local repo (since we're in a hurry). It's later separately ingested into the central repo, so we don't _strictly_ need to transfer the `exposure` dimension in production, but it's here because our testbeds can't emulate that step. Visit definition is, as far as I know, done _only_ here in the local repo.

Reply:
All three of the other `visit` dimensions have a [`populated_by: visit` line](https://github.com/lsst/daf_butler/blob/main/python/lsst/daf/butler/configs/dimensions.yaml#L430) in their configs, but AFAICT it's not actually _used_ anywhere in `daf_butler`.

Reply:
This does look exactly like the use case for [DM-34838](https://jira.lsstcorp.org/browse/DM-34838) which is about making those `populated_by` entries actually do something.  And that means this is indeed going to leave out all those auxiliary dimension records.  For now I think the list of dimension elements to export needs to be expanded to also include:

- `visit_definition`
- `visit_detector_region`
- `visit_system`
- `visit_system_membership`

And you're right that this has gone unnoticed because in the absence of those transfers it would just be future queries against the central repo that would be affected.